src/wallet/silentpayment.cpp

```diff
@@ -0,0 +1,209 @@
+    size_t len;
+    unsigned char sender_serialized_output_pubkey[33];
+    len = sizeof(sender_serialized_output_pubkey);
+    return_val = secp256k1_ec_pubkey_serialize(m_context, sender_serialized_output_pubkey, &len, &output_pubkey, SECP256K1_EC_COMPRESSED);
+    assert(return_val);
+
+    CPubKey pubKey = CPubKey(sender_serialized_output_pubkey);
+
+    assert(pubKey.IsFullyValid());
+
+    return XOnlyPubKey(pubKey);
+}
+
+CKey Sender::SumPrivateKeys(const std::vector<CKey>& sender_secret_keys)
+{
+    auto context = secp256k1_context_create(SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY);
+
+    assert(!sender_secret_keys.empty());
+
+    const auto& first_seckey{sender_secret_keys.at(0)};
+
+    unsigned char seckey[32];
+    memcpy(seckey, first_seckey.data(), 32);
+    assert(memcmp(seckey, first_seckey.data(), 32) == 0);
+
+    if (first_seckey.GetPubKey()[0] == 3) {
+        int return_val = secp256k1_ec_seckey_negate(context, seckey);
+        assert(return_val);
+    }
+
+    if (sender_secret_keys.size() > 1) {
+        for (size_t i = 1; i < sender_secret_keys.size(); i++) {
+
+            const auto& sender_seckey{sender_secret_keys.at(i)};
+
+            unsigned char seckey_i[32];
+            memcpy(seckey_i, sender_seckey.begin(), 32);
+            assert(memcmp(seckey_i, sender_seckey.begin(), 32) == 0);
+
+            if (sender_seckey.GetPubKey()[0] == 3) {
+                int return_val = secp256k1_ec_seckey_negate(context, seckey_i);
+                assert(return_val);
+            }
+
+            int return_val = secp256k1_ec_seckey_tweak_add(context, seckey, seckey_i);
+            assert(return_val);
+        }
+    }
+
+
+
+    CKey ckey;
+    ckey.Set(std::begin(seckey), std::end(seckey), true);
+
+    secp256k1_context_destroy(context);
+    memset(seckey, 0, sizeof(seckey));
+
+    return ckey;
+}
+
+Recipient::Recipient(const CKey& recipient_secret_key, const XOnlyPubKey& sender_x_only_public_key)
+{
+    m_context = secp256k1_context_create(SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY);
+
+    int return_val = secp256k1_keypair_create(m_context, &m_recipient_keypair, recipient_secret_key.begin());
+    assert(return_val);
+
+    secp256k1_xonly_pubkey parsed_sender_x_only_public_key;
+    return_val = secp256k1_xonly_pubkey_parse(m_context, &parsed_sender_x_only_public_key, sender_x_only_public_key.data());
+    assert(return_val);
+
+    return_val = secp256k1_ecdh_xonly(m_context, m_shared_secret, &parsed_sender_x_only_public_key, recipient_secret_key.begin(), nullptr, nullptr);
+    assert(return_val);
+
+    return_val = secp256k1_ec_seckey_verify(secp256k1_context_no_precomp, m_shared_secret);
+    assert(return_val);
+}
+
+Recipient::~Recipient()
+{
+    secp256k1_context_destroy(m_context);
+    memset(m_shared_secret, 0, sizeof(m_shared_secret));
+    memset(m_recipient_keypair.data, 0, sizeof(m_recipient_keypair.data));
+}
+
+CKey Recipient::Tweak(const int32_t& identifier)
```

Comment:
60537b92b0f27cab91fbb61da0d195197146bff6
I believe this can be made `const`

Reply:
Done in https://github.com/bitcoin/bitcoin/commit/fb014fefff60d87cf21f220fa756aa449bd6daf4.